infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,327 @@
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.replication.fsm.IllegalTransitionException;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationEvent;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationFSM;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationState;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationStateType;
+import org.corfudb.infrastructure.logreplication.runtime.CorfuLogReplicationRuntime;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.IllegalRuntimeTransitionException;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeState;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.sink.LogReplicationSinkEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.sink.RemoteSourceLeadershipManager;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * This class is a singleton that manages the tasks submitted by the runtime FSM, the replication FSM and also for the
+ * sink tasks when SINK is the connection starter. All sessions share the same thread pool per fsm type.
+ *
+ * The order of processing the enqueued events, for a given session, is guaranteed by maintaining a list of incoming
+ * events for session. Any new events get appended to the list.
+ * When a thread is assigned a task, we check against this list. If the task appears at index 0, the thread is allowed
+ * to process the event, otherwise, the thread waits until its notified by another thread processing the event at index 0.
+ */
+@Slf4j
+public class FsmTaskManager {
+
+    private ScheduledExecutorService runtimeWorker = null;
+    private ScheduledExecutorService replicationWorker = null;
+    private ScheduledExecutorService sinkTaskWorker = null;
+
+    // This data structure is used to maintain the order of runtime fsm events processed for a given session.
+    //
+    // The key is a sessionName instead of the session because protobuf's hashcode is not consistent. If a same session is
+    // stopped and started, the hashcode may change. Since we synchronize within a session to guarantee the order of
+    // events being processed, the inconsistency of the hashcode can block a thread for ever causing the fsm to freeze.
```

Comment:
The hashcode inconsistency is addressed only during the creation the sessoinName. The `hashcode()` of the session is not overridden as its not possible to do so with protobuf messages.
hence  thought of explaining as to why the session object is not used as the key because it does seem natural to use the it instead of the sessionName.
Please let me know if its necessary.